From 857e8627a9fa53a5035723184b6f7bf816c88278 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:28:02 -0400
Subject: [tools] perf: Synthesize anon MMAP records again

Message-id: <1408462094-14194-314-git-send-email-jolsa@redhat.com>
Patchwork-id: 88252
O-Subject: [PATCH RHEL7.1 BZ1131394 313/325] perf tools: Synthesize anon MMAP records again
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit 9d4ecc8893832337daf241236841db966fa53489
Author: Don Zickus <dzickus@redhat.com>
Date: Wed Nov 13 15:32:06 2013 -0300

description
===========
When introducing the PERF_RECORD_MMAP2 in:

5c5e854bc760 perf tools: Add attr->mmap2 support

A check for the number of entries parsed by sscanf was introduced that
assumed all of the 8 fields needed to be correctly parsed so that
particular /proc/pid/maps line would be considered synthesizable.

That broke anon records synthesizing, as it doesn't have the 'execname'
field.

Fix it by keeping the sscanf return check, changing it to not require
that the 'execname' variable be parsed, so that the preexisting logic
can kick in and set it to '//anon'.

This should get things like JIT profiling working again.

[ commit log message is mine, dzickus reported the problem with a patch ]
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/util/event.c b/tools/perf/util/event.c
index 6e3a846..bb788c1 100644
--- a/tools/perf/util/event.c
+++ b/tools/perf/util/event.c
@@ -209,8 +209,10 @@ static int perf_event__synthesize_mmap_events(struct perf_tool *tool,
          &event->mmap.start, &event->mmap.len, prot,
          &event->mmap.pgoff,
          execname);
-
-  if (n != 5)
+  /*
+    * Anon maps don't have the execname.
+    */
+  if (n < 4)
    continue;
   /*
    * Just like the kernel, see __perf_event_mmap in kernel/perf_event.c
-- 
1.7.1